packages/amplify-graphql-relational-transformer/src/utils.ts

```diff
@@ -193,25 +209,157 @@ export const ensureReferencesArray = (
+    if (getBaseType(associatedField.type) !== object.name.value) {
+      return [];
+    }
+
+    // associated field has a counterpart directive as defined in {@link getAssociatedRelationalDirectiveTypes}
+    const associatedRelationalDirective = associatedField.directives?.find((directive) => {
+      return getAssociatedRelationalDirectiveTypes(config.directiveName)
+        .map((directiveType) => directiveType)
+        .includes(directive.name.value);
+    });
+
+    if (!associatedRelationalDirective) {
+      throw new InvalidDirectiveError(`Uni-directional relationships are not supported. ${expectedBidirectionalErrorMessages()}`);
+    }
+
+    // extract the `references` arguments
+    const associatedDirectiveReferencesArgNode = associatedRelationalDirective.arguments?.find((arg) => arg.name.value === 'references')
+      ?.value as ListValueNode | StringValueNode;
+
+    const getReferencesFromArgNode = (argument: ListValueNode | StringValueNode): string[] => {
+      if (argument.kind === 'ListValue') {
+        return argument.values.map((value) => (value as StringValueNode).value);
+      }
+      return [argument.value];
+    };
+    const associatedReferences = getReferencesFromArgNode(associatedDirectiveReferencesArgNode);
+
+    return [
+      {
+        associatedField,
+        associatedReferences,
+      },
+    ];
+  });
+
+  // if we didn't find any matches, bail. This is not a bidirectional relationship.
+  // the `length > 1` is validated further upstream of relational transformers.
+  if (!associatedConnection || associatedConnection.length === 0) {
+    throw new InvalidDirectiveError(`Uni-directional relationships are not supported. ${expectedBidirectionalErrorMessages()}`);
   }
 
-  /*
-    1. find related directives:
-      - if hasMany --> belongsTo
-      - if hasOne --> belongsTo
-      - if belongsTo --> hasMany | hasOne
-    2. find matching propertyName
-      - directive.arguments[].references === config.references (order matters)
-    3. validate matching type
-      - field.type.name.value === config.type
-
-  */
+  return associatedConnection[0];
+};
+
+/**
+ * Compare the types of two {@link FieldDefinitionNode}s for equality for references based relationships as defined below:
+ * - Each field's {@link TypeNode} must be either (but don't need to match). {@link ListType} will fail this check.
+ *    - {@link NamedType}
+ *    - {@link NonNullType} where the underlying type is {@link NamedType}
+ * - Each underlying {@link NamedType}.name.value must be equal **or** both must be either `ID` or `String`.
+ *
+ * Note: argument order doesn't matter.
+ * @param a {@link FieldDefinitionNode}
+ * @param b {@link FieldDefinitionNode}
+ * @returns `boolean` indicating if field types match.
+ */
+const referenceFieldTypeMatchesPrimaryKey = (a: FieldDefinitionNode, b: FieldDefinitionNode): boolean => {
+  // `String` and `ID` are considered equal types for when comparing
```

Comment:
ID is an opaque type. Should we also allow comparisons to other types?

Reply:
Discussed offline, adding here for future reference:
> ID
> A unique identifier for an object. This scalar is serialized like a String but isn’t meant to be human-readable.

[[source](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html)]

We also use `String` as the attribute type of an `ID` in DynamoDB tables.